### PR TITLE
WIP: Error Handling

### PR DESCRIPTION
**Pull request by TejasQ**
Travis Status

## Why
In our UIs that use this client, there is currently a need for error handling – the most clear case being the need to handle authentication: if a request comes back with a `401`, _do something_. By extension, if a request comes back with any non-okay status code, _do something_. This pull request aims to build something reusable and robust that will meet this need.

## In this PR
- A proposed error handling solution (more details below)
- An update to `Client`'s configuration: `requestOptions` that a user can specify for all `fetch(url, requestOptions)` requests, such as `credentials: "include"`.
- Fix for #11: CI checks with Travis integration.
- Improved TSLint settings.

## How it works
Basically, the flow is as so:
- `import` a `makeErrorHandler` factory that takes in a `Map` of key/value pairings of status codes and handler functions.
  - I understand `Map`s are not as common as POJOs, meaning that this may not be very idiomatic, but a nice benefit is that it is not prototype-linked to Object and so key-collision can't happen, along with the other benefits of being unlinked. _If_ it is more helpful, we can refactor this to be an object quite easily.
- This function gives you an `errorHandler` that takes a `Promise<Response>`.
  - If the promise resolves and the underlying `Response` has an `ok` property `true`, it just hands the same thing, a `Promise<Response>` back and it is allowed to be consumed at the wish of the developer.
  - If the promise resolves and the underlying `Response` _does not_ have an `ok` property `true`, it rejects the promise invoking the handler. The handler ought to `return` a rejection reason after any other error-handling logic.

I've tested it out so far in the Labs UI and it works quite nicely: allowing global-level request error handling with `401`s, but also local-level request error handling with `422` (for example, when bundle edits have incorrect data).

Of course, I'd like it to be improved so reviews are welcome. Ideally, it would be nice to factor this out into its own tool that can be used anywhere with any old `fetch` call: 

```js
import {
    makeErrorHandler
} from "contiamo-http-error-handler" // or something

const handle401 = (response) => {

    // Maybe there's a backup cookie somewhere...
    if (getCookie('backup-cookie')) {
        // You can even do a new fetch in response to an error!
        return fetch(response.url, {
            headers: new Headers({
                cookie: getCookie('backup-cookie')
            })
        })
    }

    // Okay, he just can't be logged in! Go to the login page.
    goToOtherView("/login")
    return "You're not logged in! Redirecting..."
}

const myExpectedErrors = new Map([
    [401, handle401],
    [404, (response) => "I CANT FIND IT!"],
    // ...
])

const handleHttpError = makeErrorHandler(myExpectedErrors)

// Later...
handleHttpError(fetch('blah'))
  .then(response => console.log('It worked!'))
  .catch(reason => console.error('It failed, here is why', reason))
```

Thoughts?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "build": "npm run package",
     "prettier": "prettier \"{./src/,./src/**/}*.ts\" --write",
     "lint": "tslint \"{./src/,./src/**/}*.ts\" --fix",
-    "precommit": "npm test && npm run prettier && npm run lint && git add .",
+    "precommit": "npm test && npm run prettier && git add .",
```
Comment (TejasQ) [inline package.json]: I don't want us or other developers to have to deal with this right now. Let's [move linting to Travis](https://github.com/Contiamo/rest-client/issues/11) later.

**Comment by micha-f** [comment]
I like the approach. 

I don't think it needs to be factored out into a separate tool right now. We already have a ton of repos to maintain.

Can you create an example of how we'd use this with the client?
- What is the default behavior when no explicit error handling is specified by the user of the client
- Is it possible to set up (some) error handlers in the configuration for the client?
- Can these and additional error handlers then be overwritten for each request (`resource.modify()` call for example)

**Comment by TejasQ** [comment]
@micha-f thanks for the feedback!  To answer your questions,
- The `errorHandler` returns a promise that is resolved or rejected. Therefore, if no explicit error handling is specified, it just falls back to a traditional `fetch()` call.
  - It may be worth considering if we ought to do something else here: warn/error/etc.
- Yes! It's _super_ composable, so _some_ handlers go in the client, and others can be specific to resources/collections as needed.
- Yes (see above point).

> Can you create an example of how we'd use this with the client?
Not sure about what you meant with this. It is currently being used with the client and can be seen in the PR. If you rather are wondering what it would look like being used with the client in a real-world scenario, I'm currently implementing this in the Labs UI to illustrate and dogfood this approach. It's working quite nicely. I will post a PR with all of that soon.

**Comment by peterszerzo** [comment]
@TejasQ I'm no good resource on back-end, but I feel like assuming that the handler can always be determined solely from the status code sounds like something that can change easily in the future. What if in some cases it is a combination of status code and some other response header? Or a field in the body? Or we need some logic to work out a different error message? In case these aren't speculative, I wonder if such a formal abstraction is necessary here. Is `if (statusCode === 401) dothis() else if (statusCode === 402) dothat()` in a single middleware really that complicated, if we can get `dothis` and `dothat` nicely separated?

**Comment by micha-f** [comment]
@peterszerzo You can always handle more cases inside of the error handlers by reflecting on other properties of the response (header etc.). 

We could support an additional catchall handler so that you do not need to handle all status codes explicitly and you can even handle unknown / unexpected ones.

File: src/error-handler.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  | 503 // Service Unavailable
+  | 504 // Gateway Timeout
+  | 505 // HTTP Version Not Supported
+
+export type ErrorMap = Map<HttpError, (r: Response) => any>
+export type ErrorHandler = (responsePromise: Promise<Response>) => Promise<Response>
+
+export function makeErrorHandler(errorMap: ErrorMap) {
+  return async (responsePromise: Promise<Response>) => {
+    const response = await responsePromise
```
Comment (micha-f) [inline src/error-handler.ts]: @TejasQ What will happen here when our network is offline? The promise will fail but have no status code?

Comment (TejasQ) [reply src/error-handler.ts]: Excellent observation. Will consider this.

**Comment by TejasQ** [comment]
Also consider timed out requests.

File: src/error-handler.ts (hunk 2 above)
Comment (TejasQ) [reply src/error-handler.ts]: After considering this further, I think the concern of this error handler ought to be confined to _only_ handling HTTP transport _errors_ and nothing else.

It seems to me like network unavailability aught to be handled not by the REST client but rather by the app implementing it in its own specific way, the same for requests that simply have no response. 

That said, the beauty of this solution is that HTTP errors are catchable – just as an offline status is catchable _as well_. So,

```js
handleHttpErrors(fetch('bleh'))
  .catch(reason => console.log('hahahah', reason))
```

would catch HTTP errors, but also catch offline errors as well just the same! It's pretty cool.

Comment (micha-f) [reply src/error-handler.ts]: What would the catch handler receive in the case of timeout and no connection?
We should at least look at how the code looks that would handle (in one function) all three of these (timeouts, no connection, our http errors)

Comment (TejasQ) [reply src/error-handler.ts]: Ideally, all promises reject with a _"rejection reason"_ [per the spec](https://promisesaplus.com/#point-55), so I'd assume that 

```js
.catch(reason => {
  // do something based on the reason...
})
```

would suffice across the board? Am I missing something? Do you have any other ideas?

Comment (micha-f) [reply src/error-handler.ts]: What do those reasons look like for timeout and network down?

Comment (TejasQ) [reply src/error-handler.ts]: Strings. In case of network down, `TypeError: Failed to fetch`.

Comment (micha-f) [reply src/error-handler.ts]: As a string?

Comment (TejasQ) [reply src/error-handler.ts]: Ah, as an `Error`.

Comment (TejasQ) [reply src/error-handler.ts]: So, `Errors` coerce to strings, but have their own constructor that gives them a `message` and a stack trace.

Comment (TejasQ) [reply src/error-handler.ts]: Interesting. It's not even an `Error`. It's an `object` but really looks like a string. UNLESS it's a `String`. Yes, that's what it is. A boxed `String`.

Comment (micha-f) [reply src/error-handler.ts]: So will it be easy to differentiate between Timeout, Connection Down and our Errors?

Comment (TejasQ) [reply src/error-handler.ts]: Yes. They all just provide wrapped strings as reasons.

Comment (TejasQ) [reply src/error-handler.ts]: image

Comment (TejasQ) [reply src/error-handler.ts]: @micha-f, that looks good actually. I'll implement that.

**Pull request merged**